File: npm/onemessagebus-sdk/scripts/pack.mjs
#!/usr/bin/env node
// Writes the publishable copy of @onemessagebus/sdk: the built `dist`, the README,
// and a manifest stamped with the workspace version — its own version, and an
// exact `onemessagebus-cli` dependency — with SDK_VERSION and CLI_VERSION stamped
// in the built `dist/version.js`. Cargo.toml's `[workspace.package]` is the one
// version source; nothing here invents one.
//
// It refuses (exit 1) rather than guess when a placeholder it stamps has moved:
// a stamp that silently missed would publish an SDK pinned to nothing.
//
// Usage: node scripts/pack.mjs [--from DIR] [--out DIR] [--version V]
//   --from     the package directory holding package.json, README.md and dist/ (this package)
//   --out      where the publishable copy goes (dist-pack/ in this package)
//   --version  the version to stamp (Cargo.toml's workspace version)
// Prints the directory written.
import { cpSync, existsSync, mkdirSync, readFileSync, rmSync, writeFileSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const PACKAGE = resolve(dirname(fileURLToPath(import.meta.url)), "..");
const ROOT = resolve(PACKAGE, "../..");
const PLACEHOLDER = "0.0.0-dev";

function fail(message, action) {
  process.stderr.write(`pack: ${message}\n  fix: ${action}\n`);
  process.exit(1);
}

function argument(name) {
  const at = process.argv.indexOf(name);
  if (at === -1) return undefined;
  const value = process.argv[at + 1];
  if (value === undefined || value.startsWith("--")) {
    fail(`${name} takes a value`, `pass ${name} <value>`);
  }
  return value;
}

function workspaceVersion() {
  const manifest = join(ROOT, "Cargo.toml");
  const text = existsSync(manifest) ? readFileSync(manifest, "utf8") : "";
  const section = /^\[workspace\.package\]\s*$([\s\S]*?)(?=^\[|(?![\s\S]))/mu.exec(text)?.[1];
  const version = section && /^version\s*=\s*"([^"]+)"/mu.exec(section)?.[1];
  if (!version) {
    fail(
      `${manifest} declares no [workspace.package] version`,
      "restore the root Cargo.toml, or pass --version",
    );
  }
  return version;
}

const from = resolve(argument("--from") ?? PACKAGE);
const out = resolve(argument("--out") ?? join(PACKAGE, "dist-pack"));
const version = argument("--version") ?? workspaceVersion();
if (!/^\d+\.\d+\.\d+(?:-[0-9A-Za-z.-]+)?$/u.test(version) || version === PLACEHOLDER) {
  fail(`${version} is not a version to publish`, "stamp a real semver version");
}

const built = join(from, "dist/version.js");
if (!existsSync(built)) {
  fail(
    `${built} does not exist`,
    "build first: `just node-sdk-dist OUT` builds and packs it in one step, or `just nx run onemessagebus-node-sdk:build` builds dist/ alone",
  );
}
const manifest = JSON.parse(readFileSync(join(from, "package.json"), "utf8"));
if (manifest.version !== PLACEHOLDER) {
  fail(
    `package.json's version is ${JSON.stringify(manifest.version)}, not the placeholder ${PLACEHOLDER} this packer stamps`,
    `set it back to ${PLACEHOLDER}; the version comes from Cargo.toml`,
  );
}
let versionModule = readFileSync(built, "utf8");
for (const name of ["SDK_VERSION", "CLI_VERSION"]) {
  const literal = new RegExp(
    `^export const ${name} = "${PLACEHOLDER.replaceAll(".", "\\.")}";$`,
    "gmu",
  );
  const found = versionModule.match(literal)?.length ?? 0;
  if (found !== 1) {
    fail(
      `${built} holds ${found} \`export const ${name} = "${PLACEHOLDER}";\` where the packer stamps exactly one`,
      `keep ${name} as that literal in src/version.ts, rebuild, and pack again`,
    );
  }
  versionModule = versionModule.replace(literal, `export const ${name} = "${version}";`);
}

manifest.version = version;
manifest.dependencies = { ...manifest.dependencies, "onemessagebus-cli": version };
delete manifest.scripts;
delete manifest.devDependencies;

rmSync(out, { recursive: true, force: true });
mkdirSync(out, { recursive: true });
cpSync(join(from, "dist"), join(out, "dist"), { recursive: true });
cpSync(join(from, "README.md"), join(out, "README.md"));
writeFileSync(join(out, "dist/version.js"), versionModule);
writeFileSync(join(out, "package.json"), `${JSON.stringify(manifest, null, 2)}\n`);
process.stdout.write(`${out}\n`);

File: python/onemessagebus-sdk/scripts/pack.py
"""Assemble a publishable copy of the SDK, every placeholder version stamped from Cargo.toml.

The package and the CLI it drives release as one version, and release-plz owns
it in the root Cargo.toml's `[workspace.package]`. The committed package holds
`0.0.0.dev0` in four places — its own version, its `onemessagebus-cli` pin, and
both constants of `_version.py` — and this script writes a copy with each
replaced, refusing when any one of them is not there exactly once: a placeholder
that moved would publish a package pinned to the wrong binary.

    bash python/onemessagebus-sdk/scripts/run python scripts/pack.py
    uv build python/onemessagebus-sdk/dist/pack/onemessagebus-sdk

It prints the directory it assembled.
"""

from __future__ import annotations

import argparse
import re
import shutil
import sys
from pathlib import Path
from typing import NoReturn

PACKAGE = Path(__file__).resolve().parents[1]
ROOT = PACKAGE.parents[1]
PLACEHOLDER = "0.0.0.dev0"
# What a publishable copy holds: the build metadata and the package source.
SHIPPED = ("pyproject.toml", "README.md", "src")


def fail(message: str) -> NoReturn:
    print(f"pack.py: {message}", file=sys.stderr)
    raise SystemExit(1)


def cargo_version(manifest: Path) -> str:
    """The `[workspace.package]` version of `manifest`."""
    section = ""
    for line in manifest.read_text(encoding="utf-8").splitlines():
        stripped = line.strip()
        if stripped.startswith("["):
            section = stripped
        elif section == "[workspace.package]" and (
            found := re.fullmatch(r'version\s*=\s*"([^"]+)"', stripped)
        ):
            return found.group(1)
    fail(f"{manifest} declares no [workspace.package] version to stamp")


def stamp(path: Path, placeholder: str, stamped: str) -> None:
    """Replace the one `placeholder` in `path`, refusing when there is not exactly one."""
    text = path.read_text(encoding="utf-8")
    found = text.count(placeholder)
    if found != 1:
        fail(
            f"{path} holds {found} copies of `{placeholder}` where exactly one placeholder "
            "belongs; the placeholder moved, so a packed copy would pin the wrong version. "
            "Restore it in the package source, then rerun pack.py."
        )
    path.write_text(text.replace(placeholder, stamped), encoding="utf-8")


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__.splitlines()[0])
    parser.add_argument("--source", type=Path, default=PACKAGE, help=argparse.SUPPRESS)
    parser.add_argument(
        "--manifest", type=Path, default=ROOT / "Cargo.toml", help=argparse.SUPPRESS
    )
    parser.add_argument(
        "--output", type=Path, default=PACKAGE / "dist" / "pack", help=argparse.SUPPRESS
    )
    args = parser.parse_args(argv)
    version = cargo_version(args.manifest)
    destination = args.output / "onemessagebus-sdk"
    shutil.rmtree(destination, ignore_errors=True)
    destination.mkdir(parents=True)
    ignore = shutil.ignore_patterns("__pycache__", "*.pyc")
    for name in SHIPPED:
        source = args.source / name
        if source.is_dir():
            shutil.copytree(source, destination / name, ignore=ignore)
        else:
            shutil.copy2(source, destination / name)
    project = destination / "pyproject.toml"
    stamp(project, f'version = "{PLACEHOLDER}"', f'version = "{version}"')
    stamp(project, f'"onemessagebus-cli=={PLACEHOLDER}"', f'"onemessagebus-cli=={version}"')
    constants = destination / "src" / "onemessagebus" / "_version.py"
    stamp(constants, f'__version__ = "{PLACEHOLDER}"', f'__version__ = "{version}"')
    stamp(constants, f'CLI_VERSION = "{PLACEHOLDER}"', f'CLI_VERSION = "{version}"')
    print(destination)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
